File: travel-agent-strands/src/utils/constraints.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

"""Constraint helpers for the route planner — port of utils/constraints.ts.

These are deterministic helpers used by ``plan_route``. The closure-detection
logic here is the baseline that Module 4 (model swap pedagogy) depends on, so
keep parity with the TypeScript implementation.
"""

from __future__ import annotations

from typing import List, Optional

from ..data.attractions import attractions
from ..data.distances import distances
from ..data.types import Attraction, ClosureConflict


DAYS_OF_WEEK: List[str] = [
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
    "Sunday",
]


def get_day_offset(start_day: str, offset: int) -> str:
    """Return the day of the week ``offset`` days from ``start_day``."""
    if start_day not in DAYS_OF_WEEK:
        return start_day
    start_index = DAYS_OF_WEEK.index(start_day)
    return DAYS_OF_WEEK[(start_index + offset) % 7]


def parse_time_to_minutes(time_str: str) -> int:
    """Parse ``HH:MM`` to total minutes since midnight."""
    hours, minutes = time_str.split(":")
    return int(hours) * 60 + int(minutes)


def find_attraction(name: str) -> Optional[Attraction]:
    """Look up an attraction by name from the mock data."""
    for a in attractions:
        if a.name == name:
            return a
    return None


def is_open_on_day(attraction: Attraction, day_of_week: str) -> bool:
    """Return True if ``attraction`` is open on ``day_of_week``."""
    return day_of_week not in attraction.closure_days


def is_open_at_time(attraction: Attraction, time_str: str) -> bool:
    """Return True if the time falls within ``[open_time, close_time)``."""
    t = parse_time_to_minutes(time_str)
    return parse_time_to_minutes(attraction.open_time) <= t < parse_time_to_minutes(
        attraction.close_time
    )


def get_travel_time(from_name: str, to_name: str) -> int:
    """Look up travel minutes between two attractions; default 60 if missing."""
    if from_name == to_name:
        return 0
    row = distances.get(from_name)
    if row is not None and to_name in row:
        return row[to_name]
    return 60


def detect_closure_conflicts(
    attraction_names: List[str], start_day: str, num_days: int
) -> List[ClosureConflict]:
    """Detect closure conflicts for each attraction on each requested day.

    Mirrors ``detectClosureConflicts`` in constraints.ts: for every (day,
    attraction) pair, if the attraction is closed that day a ``ClosureConflict``
    is appended with a suggestion pointing at the next available day.
    """
    conflicts: List[ClosureConflict] = []

    for day_offset in range(num_days):
        current_day = get_day_offset(start_day, day_offset)
        for name in attraction_names:
            attr = find_attraction(name)
            if attr is None:
                continue
            if not is_open_on_day(attr, current_day):
                # Find a suggestion: the next available day
                suggestion_day = ""
                for next_offset in range(1, 8):
                    candidate_day = get_day_offset(current_day, next_offset)
                    if is_open_on_day(attr, candidate_day):
                        suggestion_day = candidate_day
                        break

                conflicts.append(
                    ClosureConflict(
                        attraction_name=name,
                        requested_day=current_day,
                        closure_days=list(attr.closure_days),
                        suggestion=(
                            f"Reschedule to {suggestion_day} when {name} is open"
                            if suggestion_day
                            else f"{name} has limited availability; check schedule"
                        ),
                    )
                )

    return conflicts

File: travel-agent/src/utils/constraints.ts
import type { Attraction, ClosureConflict } from "../data/types.js";
import { distances } from "../data/distances.js";
import { attractions } from "../data/attractions.js";

const DAYS_OF_WEEK = [
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
  "Sunday",
];

/**
 * Get the day of the week N days from a start day.
 * e.g., getDayOffset("Monday", 1) => "Tuesday"
 */
function getDayOffset(startDay: string, offset: number): string {
  const startIndex = DAYS_OF_WEEK.indexOf(startDay);
  if (startIndex === -1) return startDay;
  return DAYS_OF_WEEK[(startIndex + offset) % 7];
}

/**
 * Parse a time string in HH:MM format to total minutes since midnight.
 */
function parseTimeToMinutes(time: string): number {
  const [hours, minutes] = time.split(":").map(Number);
  return hours * 60 + minutes;
}

/**
 * Look up an attraction by name from the mock data.
 */
function findAttraction(name: string): Attraction | undefined {
  return attractions.find((a) => a.name === name);
}

/**
 * Check if an attraction is open on a given day of the week.
 * Returns true if the attraction's closureDays does NOT include the given day.
 */
export function isOpenOnDay(attraction: Attraction, dayOfWeek: string): boolean {
  return !attraction.closureDays.includes(dayOfWeek);
}

/**
 * Check if an attraction is open at a specific time (HH:MM format).
 * Returns true if the time falls within [openTime, closeTime).
 */
export function isOpenAtTime(attraction: Attraction, time: string): boolean {
  const timeMinutes = parseTimeToMinutes(time);
  const openMinutes = parseTimeToMinutes(attraction.openTime);
  const closeMinutes = parseTimeToMinutes(attraction.closeTime);
  return timeMinutes >= openMinutes && timeMinutes < closeMinutes;
}

/**
 * Get travel time between two attractions from the distance matrix.
 * Returns 0 if same attraction, or a large default (60) if not found.
 */
export function getTravelTime(from: string, to: string): number {
  if (from === to) return 0;
  if (distances[from] && distances[from][to] !== undefined) {
    return distances[from][to];
  }
  // Large default if pair not found in matrix
  return 60;
}

/**
 * Calculate total time needed to visit a set of attractions in order.
 * Sums visit durations for all attractions in the route, plus travel times
 * between consecutive pairs.
 */
export function calculateTotalTime(
  attractionNames: string[],
  orderedRoute: string[]
): number {
  let totalMinutes = 0;

  for (let i = 0; i < orderedRoute.length; i++) {
    const attraction = findAttraction(orderedRoute[i]);
    if (attraction) {
      totalMinutes += attraction.visitDurationMinutes;
    }

    // Add travel time to next attraction
    if (i < orderedRoute.length - 1) {
      totalMinutes += getTravelTime(orderedRoute[i], orderedRoute[i + 1]);
    }
  }

  return totalMinutes;
}

/**
 * Check if a set of attractions fits within a daily time budget.
 * Calculates total time vs available minutes (endTime - startTime).
 * Returns an object with fit status, total minutes, and budget minutes.
 */
export function fitsInDailyBudget(
  orderedRoute: string[],
  startTime: string,
  endTime: string
): { fits: boolean; totalMinutes: number; budgetMinutes: number } {
  const budgetMinutes = parseTimeToMinutes(endTime) - parseTimeToMinutes(startTime);
  const totalMinutes = calculateTotalTime(orderedRoute, orderedRoute);

  return {
    fits: totalMinutes <= budgetMinutes,
    totalMinutes,
    budgetMinutes,
  };
}

/**
 * Detect all closure conflicts for attractions on given days.
 * For each day (starting from startDay for numDays), check each attraction.
 * If closed on that day, create a ClosureConflict with suggestion to reschedule.
 */
export function detectClosureConflicts(
  attractionNames: string[],
  startDay: string,
  numDays: number
): ClosureConflict[] {
  const conflicts: ClosureConflict[] = [];

  for (let dayOffset = 0; dayOffset < numDays; dayOffset++) {
    const currentDay = getDayOffset(startDay, dayOffset);

    for (const name of attractionNames) {
      const attraction = findAttraction(name);
      if (!attraction) continue;

      if (!isOpenOnDay(attraction, currentDay)) {
        // Find a suggestion: the next available day
        let suggestionDay = "";
        for (let nextOffset = 1; nextOffset <= 7; nextOffset++) {
          const candidateDay = getDayOffset(currentDay, nextOffset);
          if (isOpenOnDay(attraction, candidateDay)) {
            suggestionDay = candidateDay;
            break;
          }
        }

        conflicts.push({
          attractionName: name,
          requestedDay: currentDay,
          closureDays: attraction.closureDays,
          suggestion: suggestionDay
            ? `Reschedule to ${suggestionDay} when ${name} is open`
            : `${name} has limited availability; check schedule`,
        });
      }
    }
  }

  return conflicts;
}

/**
 * Greedy nearest-neighbor route optimization respecting opening hours.
 * Starts from the first attraction, always picks the nearest unvisited
 * attraction that is reachable within the time window.
 */
export function optimizeRoute(
  attractionNames: string[],
  startTime: string,
  endTime: string
): string[] {
  if (attractionNames.length === 0) return [];

  const endMinutes = parseTimeToMinutes(endTime);
  let currentTime = parseTimeToMinutes(startTime);
  const visited: string[] = [];
  const unvisited = [...attractionNames];

  // Start with the first attraction if it's reachable and open
  const firstAttraction = findAttraction(unvisited[0]);
  if (firstAttraction && currentTime < endMinutes) {
    visited.push(unvisited[0]);
    currentTime += firstAttraction.visitDurationMinutes;
    unvisited.splice(0, 1);
  } else if (unvisited.length > 0) {
    // If first isn't available, just remove it and try nearest-neighbor from scratch
    // Actually let's handle this in the main loop
  }

  // If first attraction couldn't be added, reset and use pure nearest-neighbor
  if (visited.length === 0 && unvisited.length > 0) {
    return [];
  }

  while (unvisited.length > 0) {
    const currentLocation = visited[visited.length - 1];
    let nearestIndex = -1;
    let nearestDistance = Infinity;

    for (let i = 0; i < unvisited.length; i++) {
      const candidateName = unvisited[i];
      const candidate = findAttraction(candidateName);
      if (!candidate) continue;

      const travelTime = getTravelTime(currentLocation, candidateName);
      const arrivalTime = currentTime + travelTime;
      const departureTime = arrivalTime + candidate.visitDurationMinutes;

      // Check if we can arrive while it's open and finish within the day's end time
      const openMinutes = parseTimeToMinutes(candidate.openTime);
      const closeMinutes = parseTimeToMinutes(candidate.closeTime);

      // Must arrive within opening hours and finish before end of day
      if (
        arrivalTime >= openMinutes &&
        arrivalTime < closeMinutes &&
        departureTime <= endMinutes
      ) {
        if (travelTime < nearestDistance) {
          nearestDistance = travelTime;
          nearestIndex = i;
        }
      }
    }

    if (nearestIndex === -1) {
      // No reachable attraction found within time constraints
      break;
    }

    const chosenName = unvisited[nearestIndex];
    const chosenAttraction = findAttraction(chosenName)!;
    const travelTime = getTravelTime(currentLocation, chosenName);

    currentTime += travelTime + chosenAttraction.visitDurationMinutes;
    visited.push(chosenName);
    unvisited.splice(nearestIndex, 1);
  }

  return visited;
}
